lvmd/command/lvm.go

```diff
@@ -415,14 +425,88 @@ func (t *ThinPool) ListVolumes() ([]*LogicalVolume, error) {
 	return ret, nil
 }
 
+// FindVolume finds a named logical volume in this thin pool
+func (t *ThinPool) FindVolume(name string) (*LogicalVolume, error) {
+	volumes, err := t.ListVolumes()
+	if err != nil {
+		return nil, err
+	}
+	for _, volume := range volumes {
+		if volume.Name() == name {
+			return volume, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
 // CreateVolume creates a thin volume from this pool.
-func (t *ThinPool) CreateVolume(name string, size uint64) (*LogicalVolume, error) {
-	if err := CallLVM("lvcreate", "-T", t.fullname, "-n", name, "-V", fmt.Sprintf("%vg", size>>30)); err != nil {
+func (t *ThinPool) CreateVolume(name string, size uint64, tags []string, stripe uint, stripeSize string, lvcreateOptions []string) (*LogicalVolume, error) {
+
+	lvcreateArgs := []string{"-T", t.FullName(), "-n", name, "-V", fmt.Sprintf("%vg", size>>30), "-W", "y", "-y"}
+	for _, tag := range tags {
+		lvcreateArgs = append(lvcreateArgs, "--addtag")
+		lvcreateArgs = append(lvcreateArgs, tag)
+	}
+	if stripe != 0 {
+		lvcreateArgs = append(lvcreateArgs, "-i", fmt.Sprintf("%d", stripe))
+
+		if stripeSize != "" {
+			lvcreateArgs = append(lvcreateArgs, "-I", stripeSize)
+		}
+	}
+	lvcreateArgs = append(lvcreateArgs, lvcreateOptions...)
+
+	if err := CallLVM("lvcreate", lvcreateArgs...); err != nil {
 		return nil, err
 	}
 	return t.vg.FindVolume(name)
 }
 
+// Free on a thinpool returns used data, metadata percentages, calculated freebytes
+// using percentages, sum of virtualsizes of all thinlvs and size of thinpool
+func (t *ThinPool) Free() (*ThinPoolUsage, error) {
+	tpu := &ThinPoolUsage{}
+	infoList, err := parseOutput("lvs", "lv_size,data_percent,metadata_percent,pool_lv", "-S", fmt.Sprintf("lv_name=%s||pool_lv=%s", t.Name(), t.Name()))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(infoList) < 1 {
+		return nil, errors.New("thin pool not found: " + t.FullName())
+	}
+
+	var virtualSize uint64
+	for _, info := range infoList {
+		if info["pool_lv"] == "" {
+			// thin pool
+			tpu.DataPercent, err = strconv.ParseFloat(info["data_percent"], 64)
+			if err != nil {
+				return nil, err
+			}
+			tpu.MetadataPercent, err = strconv.ParseFloat(info["metadata_percent"], 64)
+			if err != nil {
+				return nil, err
+			}
+			tpu.SizeBytes, err = strconv.ParseUint(info["lv_size"], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// thin lv
+			lvSize, err := strconv.ParseUint(info["lv_size"], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			virtualSize += lvSize
+		}
+	}
+
+	free := 1.0 - (tpu.DataPercent+tpu.MetadataPercent)/100
```

Comment:
Why is the metadata percent part of the calculation? It should only use the datapercent.  
Consider if a metadata pool is 15% full but the data is only 1% full.
Here the calculated free values are incorrect.

Reply:
- it depends on the context where this `free` is used in the rest of the code
1. this isn't used in scheduling decisions
2. used in monitoring however the user for thinpool should rely on data and metadata percent that are exposed when the metrics is for thinpool subsystem
3. in general topolvm freebytes on a volume group is also part of the metrics using the same notion, freebytes on thinpool is calculated based on data and metadata percent occupied in the thinpool
4. why will the free values here be incorrect as far as the calculation is concerned, as for the context user isn't going to create alerts based on this field rather the user will use data and metadata percentages

Reply:
I misunderstand too. `DataPercent` and `MetadataPercent`  may describe occupancy in different areas in an LV. In the following situation, `free` is incorrect. If you write data to the thin volume fully, you will get the following outputs.

```
$ sudo lvs
  LV        VG        Attr       LSize   Pool   Origin Data%  Meta%  Move Log Cpy%Sync Convert
  mylv      loop      Vwi-aotz--   1.00t mypool        0.01
  mypool    loop      twi-aotzD- 100.00m               100.00 12.11
  ubuntu-lv ubuntu-vg -wi-ao---- 110.00g
```

Reply:
> * it depends on the context where this `free` is used in the rest of the code
So where exactly is it used?
> 
> 1. this isn't used in scheduling decisions
> 2. used in monitoring however the user for thinpool should rely on data and metadata percent that are exposed when the metrics is for thinpool subsystem
> 3. in general topolvm freebytes on a volume group is also part of the metrics using the same notion, freebytes on thinpool is calculated based on data and metadata percent occupied in the thinpool
> 4. why will the free values here be incorrect as far as the calculation is concerned, as for the context user isn't going to create alerts based on this field rather the user will use data and metadata percentages

Reply:
The tpu.FreeBytes is using this value for its calculation. When is this used?

Reply:
- understood! thanks for pointing this out
- I'll remove this field as there was a TODO on this field which was asking whether this can be dropped altogether

Reply:
@nbalacha @llamerada-jp @cupnes since rest of the PR is reviewed multiple times, providing below diff for the changes done for the force push after removing tpu.FreeBytes, hope it'll be helpful

``` diff
diff --git a/lvmd/command/lvm.go b/lvmd/command/lvm.go
index 85a4d5ca..0d030191 100644
--- a/lvmd/command/lvm.go
+++ b/lvmd/command/lvm.go
@@ -4,7 +4,6 @@ import (
 	"errors"
 	"fmt"
 	"io"
-	"math"
 	"os"
 	"os/exec"
 	"path"
@@ -360,7 +359,6 @@ type ThinPoolUsage struct {
 	DataPercent     float64
 	MetadataPercent float64
 	VirtualBytes    uint64
-	FreeBytes       uint64
 	SizeBytes       uint64
 }
 
@@ -462,8 +460,8 @@ func (t *ThinPool) CreateVolume(name string, size uint64, tags []string, stripe
 	return t.vg.FindVolume(name)
 }
 
-// Free on a thinpool returns used data, metadata percentages, calculated freebytes
-// using percentages, sum of virtualsizes of all thinlvs and size of thinpool
+// Free on a thinpool returns used data, metadata percentages,
+// sum of virtualsizes of all thinlvs and size of thinpool
 func (t *ThinPool) Free() (*ThinPoolUsage, error) {
 	tpu := &ThinPoolUsage{}
 	infoList, err := parseOutput("lvs", "lv_size,data_percent,metadata_percent,pool_lv", "-S", fmt.Sprintf("lv_name=%s||pool_lv=%s", t.Name(), t.Name()))
@@ -501,8 +499,6 @@ func (t *ThinPool) Free() (*ThinPoolUsage, error) {
 		}
 	}
 
-	free := 1.0 - (tpu.DataPercent+tpu.MetadataPercent)/100
-	tpu.FreeBytes = uint64(math.Floor(float64(free) * float64(tpu.SizeBytes)))
 	tpu.VirtualBytes = virtualSize
 	return tpu, nil
 }
diff --git a/lvmd/vgservice.go b/lvmd/vgservice.go
index 457644e9..da2ca18c 100644
--- a/lvmd/vgservice.go
+++ b/lvmd/vgservice.go
@@ -200,7 +200,6 @@ func (s *vgService) send(server proto.VGService_WatchServer) error {
 			// include thinpoolitem in the response
 			res.Items = append(res.Items, &proto.WatchItem{
 				DeviceClass: dc.Name,
-				FreeBytes:   tpu.FreeBytes,
 				SizeBytes:   tpu.SizeBytes,
 				ThinPool:    tpi,
 			})
diff --git a/runners/metrics_exporter.go b/runners/metrics_exporter.go
index 8888bb4e..9d6e9b97 100644
--- a/runners/metrics_exporter.go
+++ b/runners/metrics_exporter.go
@@ -41,10 +41,9 @@ type NodeMetrics struct {
 
 // thinPoolMetricsExporter is the subset of metricsExporter corresponding to the deviceclass target
 type thinPoolMetricsExporter struct {
-	tpAvailableBytes *prometheus.GaugeVec
-	tpSizeBytes      *prometheus.GaugeVec
-	dataPercent      *prometheus.GaugeVec
-	metadataPercent  *prometheus.GaugeVec
+	tpSizeBytes     *prometheus.GaugeVec
+	dataPercent     *prometheus.GaugeVec
+	metadataPercent *prometheus.GaugeVec
 }
 
 type metricsExporter struct {
@@ -82,15 +81,6 @@ func NewMetricsExporter(conn *grpc.ClientConn, mgr manager.Manager, nodeName str
 	metrics.Registry.MustRegister(sizeBytes)
 
 	// metrics available under thinpool subsystem
-	tpAvailableBytes := prometheus.NewGaugeVec(prometheus.GaugeOpts{
-		Namespace:   metricsNamespace,
-		Subsystem:   "thinpool",
-		Name:        "available_bytes",
-		Help:        "LVM VG Thin Pool raw available bytes",
-		ConstLabels: prometheus.Labels{"node": nodeName},
-	}, []string{"device_class"})
-	metrics.Registry.MustRegister(tpAvailableBytes)
-
 	tpSizeBytes := prometheus.NewGaugeVec(prometheus.GaugeOpts{
 		Namespace:   metricsNamespace,
 		Subsystem:   "thinpool",
@@ -125,10 +115,9 @@ func NewMetricsExporter(conn *grpc.ClientConn, mgr manager.Manager, nodeName str
 		availableBytes: availableBytes,
 		sizeBytes:      sizeBytes,
 		thinPool: &thinPoolMetricsExporter{
-			tpAvailableBytes: tpAvailableBytes,
-			tpSizeBytes:      tpSizeBytes,
-			dataPercent:      dataPercent,
-			metadataPercent:  metadataPercent,
+			tpSizeBytes:     tpSizeBytes,
+			dataPercent:     dataPercent,
+			metadataPercent: metadataPercent,
 		},
 	}
 }
@@ -148,7 +137,6 @@ func (m *metricsExporter) Start(ctx context.Context) error {
 					m.sizeBytes.WithLabelValues(met.DeviceClass).Set(float64(met.SizeBytes))
 				} else if met.DeviceClassType == TypeThin {
 					// metrics for thinpool subsystem
-					m.thinPool.tpAvailableBytes.WithLabelValues(met.DeviceClass).Set(float64(met.FreeBytes))
 					m.thinPool.tpSizeBytes.WithLabelValues(met.DeviceClass).Set(float64(met.SizeBytes))
 					m.thinPool.dataPercent.WithLabelValues(met.DeviceClass).Set(float64(met.DataPercent))
 					m.thinPool.metadataPercent.WithLabelValues(met.DeviceClass).Set(float64(met.MetadataPercent))
@@ -186,7 +174,6 @@ func (m *metricsExporter) updateNode(ctx context.Context, wc proto.VGService_Wat
 			if item.ThinPool != nil {
 				ch <- NodeMetrics{
 					DeviceClass:     item.DeviceClass,
-					FreeBytes:       item.FreeBytes,
 					SizeBytes:       item.SizeBytes,
 					DataPercent:     item.ThinPool.DataPercent,
 					MetadataPercent: item.ThinPool.MetadataPercent,
@@ -227,9 +214,11 @@ func (m *metricsExporter) updateNode(ctx context.Context, wc proto.VGService_Wat
 
 		node2.Annotations[topolvm.CapacityKeyPrefix+topolvm.DefaultDeviceClassAnnotationName] = strconv.FormatUint(res.FreeBytes, 10)
 		for _, item := range res.Items {
-			freeSize := item.FreeBytes
+			var freeSize uint64
 			if item.ThinPool != nil {
 				freeSize = item.ThinPool.OverprovisionBytes
+			} else {
+				freeSize = item.FreeBytes
 			}
 			node2.Annotations[topolvm.CapacityKeyPrefix+item.DeviceClass] = strconv.FormatUint(freeSize, 10)
 		}


```